src/coreclr/jit/hwintrinsicarm64.cpp

```diff
@@ -874,6 +1057,164 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
+            if (op1->OperIs(GT_CAST))
+            {
+                // Although the API specifies a pointer, if what we have is a BYREF, that's what
+                // we really want, so throw away the cast.
+                if (op1->gtGetOp1()->TypeGet() == TYP_BYREF)
+                {
+                    op1 = op1->gtGetOp1();
+                }
+            }
+
+            NamedIntrinsic loadIntrinsic = NI_Illegal;
+
+            if (simdSize == 16)
+            {
+                loadIntrinsic = NI_AdvSimd_LoadVector128;
+            }
+            else
+            {
+                loadIntrinsic = NI_AdvSimd_LoadVector64;
+            }
+
+            retNode = gtNewSimdHWIntrinsicNode(retType, op1, loadIntrinsic, simdBaseJitType, simdSize);
+            break;
+        }
+
+        case NI_Vector64_LoadAlignedNonTemporal:
+        case NI_Vector128_LoadAlignedNonTemporal:
+        {
+            assert(sig->numArgs == 1);
+
+            if (!opts.MinOpts())
+            {
+                // ARM64 doesn't have aligned loads, but aligned loads are only validated to be
+                // aligned during minopts, so only skip the intrinsic handling if we're minopts
+                break;
+            }
+
+            op1 = impPopStack().val;
+
+            if (op1->OperIs(GT_CAST))
+            {
+                // Although the API specifies a pointer, if what we have is a BYREF, that's what
+                // we really want, so throw away the cast.
+                if (op1->gtGetOp1()->TypeGet() == TYP_BYREF)
+                {
+                    op1 = op1->gtGetOp1();
+                }
+            }
+
+            // ARM64 has non-temporal loads (LDNP) but we don't currently support them
+
+            NamedIntrinsic loadIntrinsic = NI_Illegal;
+
+            if (simdSize == 16)
+            {
+                loadIntrinsic = NI_AdvSimd_LoadVector128;
+            }
+            else
+            {
+                loadIntrinsic = NI_AdvSimd_LoadVector64;
+            }
+
+            retNode = gtNewSimdHWIntrinsicNode(retType, op1, loadIntrinsic, simdBaseJitType, simdSize);
+            break;
+        }
+
+        case NI_Vector64_LoadUnsafe:
+        case NI_Vector128_LoadUnsafe:
+        {
+            if (sig->numArgs == 2)
+            {
+                op2 = impPopStack().val;
+            }
+            else
+            {
+                assert(sig->numArgs == 1);
+            }
+
+            op1 = impPopStack().val;
+
+            if (op1->OperIs(GT_CAST))
+            {
+                // Although the API specifies a pointer, if what we have is a BYREF, that's what
+                // we really want, so throw away the cast.
+                if (op1->gtGetOp1()->TypeGet() == TYP_BYREF)
+                {
+                    op1 = op1->gtGetOp1();
+                }
+            }
+
+            if (sig->numArgs == 2)
+            {
+                op3 = gtNewIconNode(genTypeSize(simdBaseType), op2->TypeGet());
```

Comment:
Is `op2` guaranteed to be of "actual" type?

Reply:
it is a `TYP_INT`, yes.